File: src/RealWorld.Application/Services/ArticleCommandService.cs
using RealWorld.Core.Entities;
using RealWorld.Core.Interfaces;

namespace RealWorld.Application.Services;

public class ArticleCommandService
{
    private readonly IArticleRepository _articleRepository;

    public ArticleCommandService(IArticleRepository articleRepository)
    {
        _articleRepository = articleRepository;
    }

    public async Task<Article> CreateArticleAsync(string title, string description, string body, List<string> tagList, User creator)
    {
        var article = new Article(title, description, body, tagList, creator.Id);
        await _articleRepository.SaveAsync(article);
        return article;
    }

    public async Task<Article> UpdateArticleAsync(Article article, string? title, string? description, string? body)
    {
        article.Update(title, description, body);
        await _articleRepository.SaveAsync(article);
        return article;
    }
}

File: services/article-service/src/main/java/io/spring/articleservice/service/ArticleCommandService.java
package io.spring.articleservice.service;

import io.spring.articleservice.domain.Article;
import io.spring.articleservice.domain.Tag;
import io.spring.articleservice.domain.User;
import io.spring.articleservice.repository.ArticleMapper;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@AllArgsConstructor
public class ArticleCommandService {

  private ArticleMapper articleMapper;

  @Transactional
  public Article createArticle(String title, String description, String body,
      java.util.List<String> tagList, User creator) {
    Article article = new Article(title, description, body,
        tagList != null ? tagList : java.util.Collections.emptyList(), creator.getId());
    articleMapper.insert(article);
    for (Tag tag : article.getTags()) {
      Tag existingTag = articleMapper.findTag(tag.getName());
      if (existingTag == null) {
        articleMapper.insertTag(tag);
      } else {
        tag.setId(existingTag.getId());
      }
      articleMapper.insertArticleTagRelation(article.getId(), tag.getId());
    }
    return article;
  }

  public Article updateArticle(Article article, String title, String description, String body) {
    article.update(title, description, body);
    articleMapper.update(article);
    return article;
  }
}
